\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={full\_CNV\_pipeline},
            pdfauthor={Lorke},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\newcommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{full\_CNV\_pipeline}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
    \author{Lorke}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
      \predate{\centering\large\emph}
  \postdate{\par}
    \date{12/2/2018}


\begin{document}
\maketitle

\subsection{Bash scripts used for tblastn
prep}\label{bash-scripts-used-for-tblastn-prep}

\begin{Shaded}
\begin{Highlighting}[]

\CommentTok{# get SRA files with raw PE reads from ENA site for each species and split them in paired fastq files}
\ExtensionTok{fastq-dump}\NormalTok{ --split-files -I reads.sra}
\CommentTok{# prep fastqc reports}
\ExtensionTok{fastqc}\NormalTok{ reads1.fq}

\CommentTok{# remove adapters where needed, Trimmomatic}
\VariableTok{FILE_PATH=}\NormalTok{/home/derezanin/temp_storage/brown_bear/raw_reads/}
\VariableTok{TRIMMOMATIC=}\NormalTok{/usr/local/bioinf/Trimmomatic-0.35/trimmomatic-0.35.jar}

\ExtensionTok{java}\NormalTok{ -jar }\VariableTok{$TRIMMOMATIC}\NormalTok{ PE \textbackslash{}}
\NormalTok{-threads 16 \textbackslash{}}
\VariableTok{$FILE_PATH}\NormalTok{/read1.fastq \textbackslash{}}
\VariableTok{$FILE_PATH}\NormalTok{/read2.fastq \textbackslash{}}
\VariableTok{$FILE_PATH}\NormalTok{/trimmed_reads/paired_read_1.fastq \textbackslash{}}
\VariableTok{$FILE_PATH}\NormalTok{/trimmed_reads/unpaired_read_1.fastq \textbackslash{}}
\VariableTok{$FILE_PATH}\NormalTok{/trimmed_reads/paired_read_2.fastq \textbackslash{}}
\VariableTok{$FILE_PATH}\NormalTok{/trimmed_reads/unpaired_read_2.fastq \textbackslash{}}
\NormalTok{ILLUMINACLIP:adapters.fa:2:30:10 \textbackslash{}}
\NormalTok{SLIDINGWINDOW:4:15 \textbackslash{}}
\NormalTok{MINLEN:100}

\CommentTok{# crop reads to length=100bp where needed, Trimmomatic(CROP:100)}

\CommentTok{# transform fastq files to fasta with:}
\ExtensionTok{fastq2fasta.sh}\NormalTok{ read1.fq }\OperatorTok{>}\NormalTok{ read1.fa}

\CommentTok{#fastq2fasta.sh:}
\CommentTok{#!/usr/bin/env bash}
\CommentTok{#INPUT_FILES=$1}
\CommentTok{#replace comma by space}
\CommentTok{#INPUT_FILES=$\{INPUT_FILES//,/ \}}
\FunctionTok{cat} \VariableTok{$@} \KeywordTok{|} \FunctionTok{awk} \StringTok{'\{if(NR%4==1 || NR%4==2) print $0 \}'} \KeywordTok{|} \FunctionTok{sed} \StringTok{'s/^@/\textbackslash{}>/'}

\CommentTok{# concatenate all PE reads.fa in one file before blastdb indexing}
\FunctionTok{cat}\NormalTok{ read1.fa read2.fa }\OperatorTok{>>}\NormalTok{ merged_reads.fa }

\CommentTok{# prep blastdb out of merged reads for each species}
\BuiltInTok{time}\NormalTok{ makeblastdb -in merged_reads.fa  -dbtype nucl -parse_seqids -out merged_readsdb}
\CommentTok{#time for bear reads: 16 - 23 hours per db}
\end{Highlighting}
\end{Shaded}

\subsection{tblastn reference gene set(query) vs.~raw
reads(db)}\label{tblastn-reference-gene-setquery-vs.raw-readsdb}

Reference gene set contains 19 files, each with consensus protein
sequence(80aa long fragment) of a gene putatively occuring in a single
copy in reference mammal species (human,cow,horse,pig,mouse).

\begin{Shaded}
\begin{Highlighting}[]

\CommentTok{# tblastn (query=protein vs. db=nucleotide seqs which get translated,so aln type is prot vs. prot in the end)}

\KeywordTok{for} \ExtensionTok{f}\NormalTok{ in /home/derezanin/species_comp/Ortho_ref_gene_set_Geneious/80aa_ref_set/*.fa}
\KeywordTok{do}
        \VariableTok{file_name=$(}\FunctionTok{basename} \VariableTok{$f)}
        \ExtensionTok{tblastn}\NormalTok{ \textbackslash{}}
\NormalTok{        -query }\VariableTok{$f}\NormalTok{ \textbackslash{}}
\NormalTok{        -db /hts/Lorena_tmp/bears/brown_bear/blastdb/trimmed_readsDB/merged_b_beardb \textbackslash{}}
\NormalTok{        -out /hts/Lorena_tmp/bears/brown_bear/tblastn_hits/e2_hits_80aa/}\VariableTok{$file_name}\StringTok{"_e2_80aa_hits"}\NormalTok{ \textbackslash{}}
\NormalTok{        -outfmt 6 \textbackslash{}}
\NormalTok{        -evalue 1e-2 \textbackslash{}}
\NormalTok{        -num_threads 16 \textbackslash{}}
        \KeywordTok{;} \KeywordTok{done}
\end{Highlighting}
\end{Shaded}

\subsection{tblastn MHC gene set(query) vs.~raw
reads(db)}\label{tblastn-mhc-gene-setquery-vs.raw-readsdb}

MHC gene set contains 4 files, each with protein sequence(80aa long
fragment) of a human MHC gene(DQA,DQB,DRA,DRB\_exon2).

\begin{Shaded}
\begin{Highlighting}[]

\KeywordTok{for} \ExtensionTok{f}\NormalTok{ in /home/derezanin/species_comp/Ortho_ref_gene_set_Geneious/80aa_MHC_set/*.fa}
\KeywordTok{do}
        \VariableTok{file_name=$(}\FunctionTok{basename} \VariableTok{$f)}
        \ExtensionTok{tblastn}\NormalTok{ \textbackslash{}}
\NormalTok{        -query }\VariableTok{$f}\NormalTok{ \textbackslash{}}
\NormalTok{        -db /hts/Lorena_tmp/bears/brown_bear/blastdb/trimmed_readsDB/merged_b_beardb \textbackslash{}}
\NormalTok{        -out /hts/Lorena_tmp/bears/brown_bear/tblastn_hits/e2_hits_80aa/}\VariableTok{$file_name}\StringTok{"_e2_80aa_MHC_hits"}\NormalTok{ \textbackslash{}}
\NormalTok{        -outfmt 6 \textbackslash{}}
\NormalTok{        -evalue 1e-2 \textbackslash{}}
\NormalTok{        -num_threads 16 \textbackslash{}}
        \KeywordTok{;} \KeywordTok{done}

\CommentTok{# tblastn output for each species is passed further for copy number estimation (check hits_table_generator chunk)}
\end{Highlighting}
\end{Shaded}

\subsection{Defining the copy number iteration
functions}\label{defining-the-copy-number-iteration-functions}

\subsection{Generating table with tblastn hits for each
species}\label{generating-table-with-tblastn-hits-for-each-species}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Create table with hits for each ref. marker for each e-value cutoff (-4, -6, -8, -10)}

\NormalTok{b.bear <-}\StringTok{ }\KeywordTok{list}\NormalTok{(}\DataTypeTok{name =} \StringTok{"brown_bear"}\NormalTok{, }\DataTypeTok{G.size =} \DecValTok{2110508336}\NormalTok{, }\DataTypeTok{N.reads =} \DecValTok{477378468}\NormalTok{)}
\NormalTok{panda <-}\StringTok{ }\KeywordTok{list}\NormalTok{(}\DataTypeTok{name =} \StringTok{"giant_panda"}\NormalTok{, }\DataTypeTok{G.size =} \DecValTok{2405352861}\NormalTok{, }\DataTypeTok{N.reads =} \DecValTok{877225626}\NormalTok{)}
\NormalTok{p.bear <-}\StringTok{ }\KeywordTok{list}\NormalTok{(}\DataTypeTok{name =} \StringTok{"polar_bear"}\NormalTok{, }\DataTypeTok{G.size =} \DecValTok{2192934624}\NormalTok{, }\DataTypeTok{N.reads =} \DecValTok{185147862}\NormalTok{)}
\NormalTok{sun_bear <-}\StringTok{ }\KeywordTok{list}\NormalTok{(}\DataTypeTok{name =} \StringTok{"sun_bear"}\NormalTok{, }\DataTypeTok{G.size =} \DecValTok{2192934796}\NormalTok{, }\DataTypeTok{N.reads =} \DecValTok{301081988}\NormalTok{)}

\NormalTok{all_species <-}\StringTok{ }\KeywordTok{list}\NormalTok{(b.bear, panda, p.bear, sun_bear)}


\NormalTok{get_species_files <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(species_name, gene_type) \{}
\NormalTok{  species_path <-}
\StringTok{    }\ControlFlowTok{if}\NormalTok{ (gene_type }\OperatorTok{==}\StringTok{ "reference"}\NormalTok{) \{}
      \KeywordTok{paste0}\NormalTok{(}\StringTok{"/home/derezanin/temp_storage/bears/"}\NormalTok{, species_name, }\StringTok{"/tblastn_hits/e2_hits_80aa"}\NormalTok{)}
\NormalTok{    \} }\ControlFlowTok{else} \ControlFlowTok{if}\NormalTok{ (gene_type }\OperatorTok{==}\StringTok{ "mhc"}\NormalTok{) \{}
      \KeywordTok{paste0}\NormalTok{(}\StringTok{"/home/derezanin/temp_storage/bears/"}\NormalTok{, species_name, }\StringTok{"/tblastn_hits/e2_MHC_hits_80aa"}\NormalTok{)}
\NormalTok{    \} }\ControlFlowTok{else}\NormalTok{ \{}
      \KeywordTok{stop}\NormalTok{(}\KeywordTok{paste}\NormalTok{(}\StringTok{"Unknown gene_type:"} \OperatorTok{+}\StringTok{ }\NormalTok{gene_type))}
\NormalTok{    \}}
\NormalTok{  files <-}\StringTok{ }\KeywordTok{list.files}\NormalTok{(}\DataTypeTok{path =}\NormalTok{ species_path, }\DataTypeTok{full.names =} \OtherTok{TRUE}\NormalTok{)}
  \KeywordTok{return}\NormalTok{(files)}
\NormalTok{\}}

\NormalTok{get_gene_e_name <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(gene_name, e) \{}
  \KeywordTok{return}\NormalTok{(}\KeywordTok{paste0}\NormalTok{(gene_name, }\StringTok{"_e_"}\NormalTok{, e))}
\NormalTok{\}}

\NormalTok{get_gene_names <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(gene_type) \{}
\NormalTok{  first_species =}\StringTok{ }\NormalTok{all_species[[}\DecValTok{1}\NormalTok{]]}
\NormalTok{  files =}\StringTok{ }\KeywordTok{get_species_files}\NormalTok{(first_species}\OperatorTok{$}\NormalTok{name, gene_type)}
\NormalTok{  gene_names <-}\StringTok{ }\KeywordTok{c}\NormalTok{()}
  \ControlFlowTok{for}\NormalTok{ ( f }\ControlFlowTok{in}\NormalTok{ files ) \{}
\NormalTok{    data <-}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(f, }\DataTypeTok{header=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{sep=}\StringTok{"}\CharTok{\textbackslash{}t}\StringTok{"}\NormalTok{)}
\NormalTok{    gene_name <-}\StringTok{ }\KeywordTok{strsplit}\NormalTok{(data[}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{], }\StringTok{"_"}\NormalTok{)[[}\DecValTok{1}\NormalTok{]][}\DecValTok{1}\NormalTok{]}
\NormalTok{    gene_names <-}\StringTok{ }\KeywordTok{c}\NormalTok{(gene_names, gene_name)}
\NormalTok{  \}}
  
  \KeywordTok{return}\NormalTok{(gene_names)}
\NormalTok{\}}

\NormalTok{initialize_hits_table <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(all_species, gene_type) \{}
  
\NormalTok{  num_columns =}\StringTok{ }\KeywordTok{length}\NormalTok{(all_species)}
\NormalTok{  gene_names <-}\StringTok{ }\KeywordTok{get_gene_names}\NormalTok{(gene_type)}
\NormalTok{  num_genes =}\StringTok{ }\KeywordTok{length}\NormalTok{(gene_names)}
\NormalTok{  e_cutoffs <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{4}\NormalTok{,}\DecValTok{6}\NormalTok{,}\DecValTok{8}\NormalTok{,}\DecValTok{10}\NormalTok{)}
\NormalTok{  num_e_cutoffs =}\StringTok{ }\KeywordTok{length}\NormalTok{(e_cutoffs)}
\NormalTok{  num_rows =}\StringTok{ }\NormalTok{num_genes }\OperatorTok{*}\StringTok{ }\NormalTok{num_e_cutoffs}

\NormalTok{  gene_e_names =}\StringTok{ }\KeywordTok{c}\NormalTok{()}
  \ControlFlowTok{for}\NormalTok{ (gene_name }\ControlFlowTok{in}\NormalTok{ gene_names) \{}
    \ControlFlowTok{for}\NormalTok{ ( e }\ControlFlowTok{in}\NormalTok{ e_cutoffs ) \{}
\NormalTok{      gene_e_name <-}\StringTok{ }\KeywordTok{get_gene_e_name}\NormalTok{(gene_name, e)}
\NormalTok{      gene_e_names <-}\StringTok{ }\KeywordTok{c}\NormalTok{(gene_e_names, gene_e_name)}
\NormalTok{    \}}
\NormalTok{  \}}
  
\NormalTok{  hits_table =}\StringTok{ }\KeywordTok{matrix}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DataTypeTok{nrow =}\NormalTok{ num_rows, }\DataTypeTok{ncol =}\NormalTok{ num_columns)}
  \KeywordTok{colnames}\NormalTok{(hits_table) <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{(all_species, }\ControlFlowTok{function}\NormalTok{(x) x[[}\StringTok{"name"}\NormalTok{]])}
  \KeywordTok{rownames}\NormalTok{(hits_table) <-}\StringTok{ }\NormalTok{gene_e_names}
  
  \KeywordTok{return}\NormalTok{(hits_table)}
\NormalTok{\}}

\NormalTok{fill_hits <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(all_species, hits, gene_type) \{}
  \ControlFlowTok{for}\NormalTok{ (s }\ControlFlowTok{in}\NormalTok{ all_species) \{}
\NormalTok{    files <-}\StringTok{ }\KeywordTok{get_species_files}\NormalTok{(s}\OperatorTok{$}\NormalTok{name, gene_type)}
    
    \ControlFlowTok{for}\NormalTok{ ( f }\ControlFlowTok{in}\NormalTok{ files ) \{}
\NormalTok{      data <-}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(f, }\DataTypeTok{header=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{sep=}\StringTok{"}\CharTok{\textbackslash{}t}\StringTok{"}\NormalTok{)}
\NormalTok{      data.e <-}\StringTok{ }\NormalTok{data[,}\DecValTok{11}\NormalTok{]}
\NormalTok{      gene_name <-}\StringTok{ }\KeywordTok{strsplit}\NormalTok{(data[}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{], }\StringTok{"_"}\NormalTok{)[[}\DecValTok{1}\NormalTok{]][}\DecValTok{1}\NormalTok{]}
      
\NormalTok{      e_cutoffs <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{4}\NormalTok{,}\DecValTok{6}\NormalTok{,}\DecValTok{8}\NormalTok{,}\DecValTok{10}\NormalTok{)}
      \ControlFlowTok{for}\NormalTok{ ( e }\ControlFlowTok{in}\NormalTok{ e_cutoffs ) \{}
\NormalTok{        num_hits <-}\StringTok{ }\KeywordTok{length}\NormalTok{(data.e[data.e}\OperatorTok{<=}\DecValTok{10}\OperatorTok{^-}\NormalTok{e])}
\NormalTok{        gene_e_name <-}\StringTok{ }\KeywordTok{get_gene_e_name}\NormalTok{(gene_name, e)}
\NormalTok{        hits[gene_e_name, s}\OperatorTok{$}\NormalTok{name] <-}\StringTok{ }\NormalTok{num_hits}
\NormalTok{      \}}
\NormalTok{    \} }
\NormalTok{  \}}
  
  \KeywordTok{return}\NormalTok{(hits)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\subsection{Script 1 of the
procedure.}\label{script-1-of-the-procedure.}

Here we first decide which reference markers are possible to use for
estimating copy numbers.

\begin{itemize}
\tightlist
\item
  Step 1 - Sanity check on read coverage.
\item
  Step 2 - Check on effect of BLAST-cutoff.
\item
  Step 3 - Final check if copy number estimates are not 0.
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{options}\NormalTok{( }\DataTypeTok{stringsAsFactors =} \OtherTok{FALSE}\NormalTok{ )}

\NormalTok{empty_ref_hits <-}\StringTok{ }\KeywordTok{initialize_hits_table}\NormalTok{(all_species, }\StringTok{"reference"}\NormalTok{)}
\NormalTok{ref_gene_names <-}\StringTok{ }\KeywordTok{get_gene_names}\NormalTok{(}\DataTypeTok{gene_type =} \StringTok{"reference"}\NormalTok{)}
\NormalTok{ref.hits <-}\StringTok{ }\KeywordTok{fill_hits}\NormalTok{(all_species, empty_ref_hits, }\DataTypeTok{gene_type =} \StringTok{"reference"}\NormalTok{)}
\NormalTok{uref <-}\StringTok{ }\KeywordTok{unique}\NormalTok{(ref_gene_names)}

\CommentTok{# marker length in bp}
\NormalTok{marker.length <-}\StringTok{ }\DecValTok{80}\OperatorTok{*}\DecValTok{3}                            
\NormalTok{G.size <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{(all_species, }\ControlFlowTok{function}\NormalTok{(x) x[[}\StringTok{"G.size"}\NormalTok{]])   }\CommentTok{# assembly size estimated from k=31-91, s=2 (kmergenie)}
\NormalTok{n.species <-}\StringTok{ }\KeywordTok{ncol}\NormalTok{( ref.hits )                       }\CommentTok{# number of species observed }
\NormalTok{n.ref <-}\StringTok{ }\KeywordTok{length}\NormalTok{( uref )                          }\CommentTok{# number of reference markers/genes}
\NormalTok{N.reads <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{(all_species, }\ControlFlowTok{function}\NormalTok{(x) x[[}\StringTok{"N.reads"}\NormalTok{]])   }\CommentTok{# N of concatenated raw PE reads (~100bp)}
\NormalTok{cuts <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{,}\OperatorTok{-}\DecValTok{6}\NormalTok{,}\OperatorTok{-}\DecValTok{8}\NormalTok{,}\OperatorTok{-}\DecValTok{10}\NormalTok{)}
\KeywordTok{cat}\NormalTok{( }\StringTok{"   we have data for"}\NormalTok{, n.ref, }\StringTok{"reference markers}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    we have data for 19 reference markers
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{cat}\NormalTok{( }\StringTok{"   we have data for"}\NormalTok{, n.species, }\StringTok{"species}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    we have data for 4 species
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{cat}\NormalTok{( }\StringTok{"   reference markers are"}\NormalTok{, marker.length, }\StringTok{"bases long}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    reference markers are 240 bases long
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.species, }\ControlFlowTok{function}\NormalTok{(i)\{}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(ref.hits)[i], }\StringTok{"has"}\NormalTok{, G.size[i],}\StringTok{"basepairs and"}\NormalTok{,N.reads[i],}\StringTok{"reads}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\} )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## brown_bear has 2110508336 basepairs and 477378468 reads
## giant_panda has 2405352861 basepairs and 877225626 reads
## polar_bear has 2192934624 basepairs and 185147862 reads
## sun_bear has 2192934796 basepairs and 301081988 reads
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{### First data sanity check - checking for too low coverage}
\NormalTok{### Some reference markers may, for some reason, have extremely low coverage}
\NormalTok{### We would like to exclude these since they may affect all other estimates}
\NormalTok{### of copy numbers later}
\NormalTok{low.cov <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\DecValTok{0}\NormalTok{, }\DataTypeTok{nrow=}\NormalTok{n.species, }\DataTypeTok{ncol=}\NormalTok{n.ref )}
\KeywordTok{rownames}\NormalTok{( low.cov ) <-}\StringTok{ }\KeywordTok{colnames}\NormalTok{( ref.hits )}
\KeywordTok{colnames}\NormalTok{( low.cov ) <-}\StringTok{ }\NormalTok{uref}
\NormalTok{coverage <-}\StringTok{ }\NormalTok{(N.reads}\OperatorTok{*}\NormalTok{marker.length)}\OperatorTok{/}\NormalTok{G.size    }\CommentTok{# This is the expected coverage per species}
\ControlFlowTok{for}\NormalTok{( i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{n.ref )\{}
  \KeywordTok{cat}\NormalTok{( }\StringTok{"Reference "}\NormalTok{, uref[i], }\StringTok{"}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
  
\NormalTok{  idd <-}\StringTok{ }\KeywordTok{seq}\NormalTok{( }\DecValTok{1}\NormalTok{, }\KeywordTok{nrow}\NormalTok{(ref.hits), }\DecValTok{4}\NormalTok{ )                          }\CommentTok{# hits under cutoff -4}
\NormalTok{  ohits_cutoff5 <-}\StringTok{ }\NormalTok{ref.hits[idd[i],]                          }\CommentTok{# observed number of hits for reference gene uref[i] at cutoff -4}
\NormalTok{  normalized_hits <-}\StringTok{ }\NormalTok{ohits_cutoff5}\OperatorTok{/}\NormalTok{coverage                   }\CommentTok{# number of hits per unit of coverage (~percentage of hits compared to coverage)}
\NormalTok{  phi1 <-}\StringTok{ }\KeywordTok{mean}\NormalTok{( normalized_hits )                             }\CommentTok{# very rough estimate of cutoff-factor Fs}
\NormalTok{  ehits1 <-}\StringTok{ }\NormalTok{coverage}\OperatorTok{*}\NormalTok{phi1                                     }\CommentTok{# expected number of hits based on phi1}
  \CommentTok{# chi- square - for diff. between observed and estimated hits is high}
\NormalTok{  rr1 <-}\StringTok{ }\KeywordTok{sign}\NormalTok{( ohits_cutoff5 }\OperatorTok{-}\StringTok{ }\NormalTok{(ehits1}\OperatorTok{-}\DecValTok{3}\OperatorTok{*}\KeywordTok{sqrt}\NormalTok{(ehits1)) )      }\CommentTok{# if rr1 is negative the coverage is very low}
 
  
\NormalTok{  idd <-}\StringTok{ }\KeywordTok{seq}\NormalTok{( }\DecValTok{2}\NormalTok{, }\KeywordTok{nrow}\NormalTok{(ref.hits), }\DecValTok{4}\NormalTok{ )                          }\CommentTok{# hits under cutoff -6}
\NormalTok{  phi2 <-}\StringTok{ }\KeywordTok{mean}\NormalTok{( ref.hits[idd[i],]}\OperatorTok{/}\NormalTok{coverage )                  }\CommentTok{# very rough estimate of cutoff-factor}
\NormalTok{  ehits2 <-}\StringTok{ }\NormalTok{coverage}\OperatorTok{*}\NormalTok{phi2                                     }\CommentTok{# expected number of hits based on phi2}
\NormalTok{  rr2 <-}\StringTok{ }\KeywordTok{sign}\NormalTok{( ref.hits[idd[i],] }\OperatorTok{-}\StringTok{ }\NormalTok{(ehits2}\OperatorTok{-}\DecValTok{3}\OperatorTok{*}\KeywordTok{sqrt}\NormalTok{(ehits2)) )  }\CommentTok{# if rr2 is negative the coverage is very low}

\NormalTok{  idd <-}\StringTok{ }\KeywordTok{seq}\NormalTok{( }\DecValTok{3}\NormalTok{, }\KeywordTok{nrow}\NormalTok{(ref.hits), }\DecValTok{4}\NormalTok{ )                          }\CommentTok{# hits under cutoff -8}
\NormalTok{  phi3 <-}\StringTok{ }\KeywordTok{mean}\NormalTok{( ref.hits[idd[i],]}\OperatorTok{/}\NormalTok{coverage )                  }
\NormalTok{  ehits3 <-}\StringTok{ }\NormalTok{coverage}\OperatorTok{*}\NormalTok{phi3                                     }\CommentTok{# expected number of hits based on phi3 }
\NormalTok{  rr3 <-}\StringTok{ }\KeywordTok{sign}\NormalTok{( ref.hits[idd[i],] }\OperatorTok{-}\StringTok{ }\NormalTok{(ehits3}\OperatorTok{-}\DecValTok{3}\OperatorTok{*}\KeywordTok{sqrt}\NormalTok{(ehits3)) )}

\NormalTok{  idd <-}\StringTok{ }\KeywordTok{seq}\NormalTok{( }\DecValTok{4}\NormalTok{, }\KeywordTok{nrow}\NormalTok{(ref.hits), }\DecValTok{4}\NormalTok{ )                          }\CommentTok{# hits under cutoff -10}
\NormalTok{  phi4 <-}\StringTok{ }\KeywordTok{mean}\NormalTok{( ref.hits[idd[i],]}\OperatorTok{/}\NormalTok{coverage )}
\NormalTok{  ehits4 <-}\StringTok{ }\NormalTok{coverage}\OperatorTok{*}\NormalTok{phi4}
\NormalTok{  rr4 <-}\StringTok{ }\KeywordTok{sign}\NormalTok{( ref.hits[idd[i],] }\OperatorTok{-}\StringTok{ }\NormalTok{(ehits4}\OperatorTok{-}\DecValTok{3}\OperatorTok{*}\KeywordTok{sqrt}\NormalTok{(ehits4)) )}
  
\NormalTok{  rmat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\KeywordTok{c}\NormalTok{(rr1,rr2,rr3,rr4), }\DataTypeTok{ncol=}\DecValTok{4}\NormalTok{, }\DataTypeTok{byrow=}\NormalTok{F )       }\CommentTok{# The rr1,...,rr4 values}
\NormalTok{  low.cov[,i] <-}\StringTok{ }\KeywordTok{rowSums}\NormalTok{( rmat )                              }\CommentTok{# summing the signs, i.e. need 3 out of 4 negative to}
\NormalTok{\}                                                             }\CommentTok{# get a negative sum}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Reference  AP1G1 
## Reference  ATP6V1A 
## Reference  CNOT2 
## Reference  EIF4A2 
## Reference  EIF4G2 
## Reference  GRIK2 
## Reference  KCTD9 
## Reference  KIF5B 
## Reference  MKLN1 
## Reference  NCKAP1 
## Reference  PAN3 
## Reference  PAPOLA 
## Reference  RICTOR 
## Reference  RNF38 
## Reference  SMAD2 
## Reference  SMARCA5 
## Reference  TLK2 
## Reference  TOP1 
## Reference  TRIM33
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ref.keep <-}\StringTok{ }\NormalTok{(low.cov }\OperatorTok{>=}\StringTok{ }\DecValTok{0}\NormalTok{ )}
\NormalTok{r.disc <-}\StringTok{ }\KeywordTok{rowSums}\NormalTok{(}\OperatorTok{!}\NormalTok{ref.keep)}
\NormalTok{f.disc <-}\StringTok{ }\KeywordTok{colSums}\NormalTok{( }\OperatorTok{!}\NormalTok{ref.keep )                                }\CommentTok{# number of species for which each ref.gene was discadred   }
\KeywordTok{cat}\NormalTok{( }\StringTok{"Number of discarded reference markers per species:}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Number of discarded reference markers per species:
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.species, }\ControlFlowTok{function}\NormalTok{(i)\{}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(ref.keep)[i], }\StringTok{"discards"}\NormalTok{,r.disc[i],}\StringTok{"reference markers}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\} )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## brown_bear discards 0 reference markers
## giant_panda discards 0 reference markers
## polar_bear discards 0 reference markers
## sun_bear discards 0 reference markers
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{cat}\NormalTok{( }\StringTok{"Number of discarded species per reference marker:}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Number of discarded species per reference marker:
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.ref, }\ControlFlowTok{function}\NormalTok{(i)\{}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(ref.keep)[i], }\StringTok{"is discarded in"}\NormalTok{,f.disc[i],}\StringTok{"species}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\} )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## AP1G1 is discarded in 0 species
## ATP6V1A is discarded in 0 species
## CNOT2 is discarded in 0 species
## EIF4A2 is discarded in 0 species
## EIF4G2 is discarded in 0 species
## GRIK2 is discarded in 0 species
## KCTD9 is discarded in 0 species
## KIF5B is discarded in 0 species
## MKLN1 is discarded in 0 species
## NCKAP1 is discarded in 0 species
## PAN3 is discarded in 0 species
## PAPOLA is discarded in 0 species
## RICTOR is discarded in 0 species
## RNF38 is discarded in 0 species
## SMAD2 is discarded in 0 species
## SMARCA5 is discarded in 0 species
## TLK2 is discarded in 0 species
## TOP1 is discarded in 0 species
## TRIM33 is discarded in 0 species
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{### Next, checking which reference genes have read-hits that}
\NormalTok{### cannot be explained by the cutoff-model.}
\NormalTok{### In short, we expect that as the blast-cutoff is made stricter, the}
\NormalTok{### number of hits should decrease in a smooth way.}
\NormalTok{### Here we look for cases where this is clearly violated}
\NormalTok{mse <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\DecValTok{0}\NormalTok{, }\DataTypeTok{nrow=}\NormalTok{n.species, }\DataTypeTok{ncol=}\NormalTok{n.ref )       }\CommentTok{# mse represents residual variance (S2g) for a marker }
\KeywordTok{colnames}\NormalTok{( mse ) <-}\StringTok{ }\NormalTok{uref}
\KeywordTok{rownames}\NormalTok{( mse ) <-}\StringTok{ }\KeywordTok{colnames}\NormalTok{( ref.hits )}
\ControlFlowTok{for}\NormalTok{( ss }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{n.species )\{}
  \KeywordTok{cat}\NormalTok{( }\KeywordTok{rownames}\NormalTok{(mse)[ss], }\StringTok{":}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
  \CommentTok{# same as C <- coverage[ss] }
\NormalTok{  C <-}\StringTok{ }\NormalTok{(N.reads[ss]}\OperatorTok{*}\NormalTok{marker.length)}\OperatorTok{/}\NormalTok{G.size[ss]     }\CommentTok{# coverage for this species}
\NormalTok{  idx.in <-}\StringTok{ }\KeywordTok{which}\NormalTok{( ref.keep[ss,] )                }\CommentTok{# indices of ref. genes used for this species}
\NormalTok{  n.in <-}\StringTok{ }\KeywordTok{length}\NormalTok{( idx.in )                        }\CommentTok{# number of ref. genes used for this species}
  \CommentTok{# all observed hits for this species, columns = ref.genes, rows = 4 cutoffs}
\NormalTok{  y.mat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( ref.hits[,ss], }\DataTypeTok{nrow=}\DecValTok{4}\NormalTok{, }\DataTypeTok{ncol=}\NormalTok{n.ref, }\DataTypeTok{byrow=}\NormalTok{F ) }
\NormalTok{  y.mat <-}\StringTok{ }\NormalTok{y.mat[,idx.in]                         }\CommentTok{# hits for kept ref. genes for this species}
\NormalTok{  lst <-}\StringTok{ }\KeywordTok{copynum.iter.iter}\NormalTok{( y.mat, C, }\KeywordTok{rep}\NormalTok{( }\DecValTok{1}\NormalTok{, n.in ), }\DataTypeTok{as.integer=}\OtherTok{FALSE}\NormalTok{ )}
  
\NormalTok{  g.mat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\KeywordTok{rep}\NormalTok{( lst}\OperatorTok{$}\NormalTok{Gamma, }\DecValTok{4}\NormalTok{ ), }\DataTypeTok{nrow=}\DecValTok{4}\NormalTok{, }\DataTypeTok{byrow=}\OtherTok{TRUE}\NormalTok{ )}
\NormalTok{  p.mat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\KeywordTok{rep}\NormalTok{( lst}\OperatorTok{$}\NormalTok{Phi, n.in ), }\DataTypeTok{ncol=}\NormalTok{n.in, }\DataTypeTok{byrow=}\OtherTok{FALSE}\NormalTok{ )}
  \CommentTok{# estimated number of ref. gene hits for this species for all 4 cutoffs}
\NormalTok{  y.hat <-}\StringTok{ }\NormalTok{C }\OperatorTok{*}\StringTok{ }\NormalTok{g.mat }\OperatorTok{*}\StringTok{ }\NormalTok{p.mat}
  \CommentTok{# difference between observed and estimated ref.gene hits }
\NormalTok{  r.mat <-}\StringTok{ }\NormalTok{y.mat }\OperatorTok{-}\StringTok{ }\NormalTok{y.hat}
  \CommentTok{# calculation of residual variance S^2g (formula 7. in SI) for each gene}
\NormalTok{  mse[ss,idx.in] <-}\StringTok{ }\KeywordTok{apply}\NormalTok{( r.mat, }\DecValTok{2}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(x)\{}\KeywordTok{sum}\NormalTok{(x}\OperatorTok{^}\DecValTok{2}\NormalTok{)}\OperatorTok{/}\NormalTok{(}\DecValTok{4}\OperatorTok{-}\DecValTok{1}\NormalTok{)\} )}
  \CommentTok{# range returns min and max of given arguments}
\NormalTok{  rr <-}\StringTok{ }\KeywordTok{range}\NormalTok{( }\KeywordTok{c}\NormalTok{( }\KeywordTok{as.vector}\NormalTok{( y.mat ), }\KeywordTok{as.vector}\NormalTok{(y.hat) ) )}
  \KeywordTok{plot}\NormalTok{( rr, rr, }\DataTypeTok{type=}\StringTok{"l"}\NormalTok{, }\DataTypeTok{col=}\StringTok{"red"}\NormalTok{, }\DataTypeTok{main=}\KeywordTok{rownames}\NormalTok{(mse)[ss], }
        \DataTypeTok{xlab=}\StringTok{"Observed number of reads"}\NormalTok{, }\DataTypeTok{ylab=}\StringTok{"Predicted number of reads"}\NormalTok{ )}
  \KeywordTok{points}\NormalTok{( y.mat, y.hat, }\DataTypeTok{pch=}\DecValTok{16}\NormalTok{ )}
  \KeywordTok{Sys.sleep}\NormalTok{( }\DecValTok{2}\NormalTok{ )}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## brown_bear :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4
\end{verbatim}

\includegraphics{full_CNV_pipeline_files/figure-latex/script1_ref_markers_mammals-1.pdf}

\begin{verbatim}
## giant_panda :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration 3 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration 3 
##  iteration.iteration 3
\end{verbatim}

\includegraphics{full_CNV_pipeline_files/figure-latex/script1_ref_markers_mammals-2.pdf}

\begin{verbatim}
## polar_bear :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration 3 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2
\end{verbatim}

\includegraphics{full_CNV_pipeline_files/figure-latex/script1_ref_markers_mammals-3.pdf}

\begin{verbatim}
## sun_bear :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 5
\end{verbatim}

\includegraphics{full_CNV_pipeline_files/figure-latex/script1_ref_markers_mammals-4.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# graph for observed and estimated ref. gene hits - closer the points are to the red line, smaller the diff.}


\NormalTok{sigma2 <-}\StringTok{ }\KeywordTok{mean}\NormalTok{( mse, }\DataTypeTok{trim=}\FloatTok{0.01}\NormalTok{ )          }\CommentTok{# sigma2 is the variance of the error term, trimmed off 1% of extreme values}
\NormalTok{lambda <-}\StringTok{ }\NormalTok{(}\DecValTok{4}\OperatorTok{-}\DecValTok{1}\NormalTok{)}\OperatorTok{*}\NormalTok{mse}\OperatorTok{/}\NormalTok{sigma2                }\CommentTok{# if lambda is large, it indicates that residual variance }
\NormalTok{limit <-}\StringTok{ }\KeywordTok{qchisq}\NormalTok{( }\FloatTok{0.99}\NormalTok{, }\DataTypeTok{df=}\NormalTok{(}\DecValTok{4}\OperatorTok{-}\DecValTok{1}\NormalTok{) )         }\CommentTok{# is much larger than we expected - data fits the model poorly}
\NormalTok{ref.keep <-}\StringTok{ }\NormalTok{(lambda }\OperatorTok{<=}\StringTok{ }\NormalTok{limit)}\OperatorTok{&}\NormalTok{ref.keep    }\CommentTok{# marker kept if lambda is lower than 99% quantile of the chi-square }
                                          \CommentTok{# distribution with 3 degrees of freedom}

\KeywordTok{cat}\NormalTok{( }\StringTok{"Number of discarded reference markers per species:}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Number of discarded reference markers per species:
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.species, }\ControlFlowTok{function}\NormalTok{(i)\{}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(ref.keep)[i], }\StringTok{"discards"}\NormalTok{,}\KeywordTok{rowSums}\NormalTok{(}\OperatorTok{!}\NormalTok{ref.keep)[i],}\StringTok{"reference markers}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\} )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## brown_bear discards 1 reference markers
## giant_panda discards 3 reference markers
## polar_bear discards 0 reference markers
## sun_bear discards 0 reference markers
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{cat}\NormalTok{( }\StringTok{"Number of discarded species per reference marker:}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Number of discarded species per reference marker:
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.ref, }\ControlFlowTok{function}\NormalTok{(i)\{}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(ref.keep)[i], }\StringTok{"is discarded in"}\NormalTok{,}\KeywordTok{colSums}\NormalTok{( }\OperatorTok{!}\NormalTok{ref.keep )[i],}\StringTok{"species}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\} )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## AP1G1 is discarded in 0 species
## ATP6V1A is discarded in 0 species
## CNOT2 is discarded in 0 species
## EIF4A2 is discarded in 0 species
## EIF4G2 is discarded in 0 species
## GRIK2 is discarded in 0 species
## KCTD9 is discarded in 0 species
## KIF5B is discarded in 1 species
## MKLN1 is discarded in 0 species
## NCKAP1 is discarded in 0 species
## PAN3 is discarded in 0 species
## PAPOLA is discarded in 0 species
## RICTOR is discarded in 0 species
## RNF38 is discarded in 0 species
## SMAD2 is discarded in 1 species
## SMARCA5 is discarded in 2 species
## TLK2 is discarded in 0 species
## TOP1 is discarded in 0 species
## TRIM33 is discarded in 0 species
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{### Third, eliminating reference genes that still estimate }
\NormalTok{### to 0 copies, as these will blow the other copy number }
\NormalTok{### estimates sky high later...}
\NormalTok{ref.hat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\DecValTok{0}\NormalTok{, }\DataTypeTok{nrow=}\NormalTok{ n.species, }\DataTypeTok{ncol=}\NormalTok{n.ref )}
\ControlFlowTok{for}\NormalTok{( ss }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{n.species)\{}
  \KeywordTok{cat}\NormalTok{( }\KeywordTok{rownames}\NormalTok{(ref.keep)[ss], }\StringTok{":}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\NormalTok{  C <-}\StringTok{ }\NormalTok{(N.reads[ss]}\OperatorTok{*}\NormalTok{marker.length)}\OperatorTok{/}\NormalTok{G.size[ss]}
\NormalTok{  idx.in <-}\StringTok{ }\KeywordTok{which}\NormalTok{( ref.keep[ss,] )}
\NormalTok{  n.in <-}\StringTok{ }\KeywordTok{length}\NormalTok{( idx.in )}
\NormalTok{  y.mat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( ref.hits[,ss], }\DataTypeTok{nrow=}\DecValTok{4}\NormalTok{, }\DataTypeTok{ncol=}\NormalTok{n.ref, }\DataTypeTok{byrow=}\NormalTok{F )}
\NormalTok{  y.mat <-}\StringTok{ }\NormalTok{y.mat[,idx.in]}
\NormalTok{  lst <-}\StringTok{ }\KeywordTok{copynum.iter.iter}\NormalTok{( y.mat, C, }\KeywordTok{rep}\NormalTok{( }\DecValTok{1}\NormalTok{, n.in ), }\DataTypeTok{as.integer=}\OtherTok{FALSE}\NormalTok{ )}
\NormalTok{  ref.hat[ss,idx.in] <-}\StringTok{ }\NormalTok{lst}\OperatorTok{$}\NormalTok{Gamma}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## brown_bear :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4 
## giant_panda :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4 
## polar_bear :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration 3 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## sun_bear :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 5
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ref.keep <-}\StringTok{ }\NormalTok{(}\KeywordTok{round}\NormalTok{(ref.hat) }\OperatorTok{>}\StringTok{ }\DecValTok{0}\NormalTok{)}\OperatorTok{&}\NormalTok{ref.keep}
\KeywordTok{cat}\NormalTok{( }\StringTok{"Number of discarded reference markers per species:}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Number of discarded reference markers per species:
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.species, }\ControlFlowTok{function}\NormalTok{(i)\{}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(ref.keep)[i], }\StringTok{"discards"}\NormalTok{,}\KeywordTok{rowSums}\NormalTok{(}\OperatorTok{!}\NormalTok{ref.keep)[i],}\StringTok{"reference markers}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\} )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## brown_bear discards 4 reference markers
## giant_panda discards 4 reference markers
## polar_bear discards 4 reference markers
## sun_bear discards 3 reference markers
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{cat}\NormalTok{( }\StringTok{"Number of species per discarded reference marker:}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Number of species per discarded reference marker:
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.ref, }\ControlFlowTok{function}\NormalTok{(i)\{}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(ref.keep)[i], }\StringTok{"is discarded in"}\NormalTok{,}\KeywordTok{colSums}\NormalTok{( }\OperatorTok{!}\NormalTok{ref.keep )[i],}\StringTok{"species}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\} )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## AP1G1 is discarded in 1 species
## ATP6V1A is discarded in 1 species
## CNOT2 is discarded in 2 species
## EIF4A2 is discarded in 0 species
## EIF4G2 is discarded in 4 species
## GRIK2 is discarded in 0 species
## KCTD9 is discarded in 0 species
## KIF5B is discarded in 1 species
## MKLN1 is discarded in 0 species
## NCKAP1 is discarded in 0 species
## PAN3 is discarded in 0 species
## PAPOLA is discarded in 0 species
## RICTOR is discarded in 0 species
## RNF38 is discarded in 0 species
## SMAD2 is discarded in 1 species
## SMARCA5 is discarded in 2 species
## TLK2 is discarded in 0 species
## TOP1 is discarded in 0 species
## TRIM33 is discarded in 3 species
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#save( ref.keep, file="res/ref_keep.RData" )}
\end{Highlighting}
\end{Shaded}

\subsection{Script 2 of the
procedure.}\label{script-2-of-the-procedure.}

Here we estimate the reference markers copy numbers.

\begin{itemize}
\tightlist
\item
  Step 1 - Point estimates
\item
  Step 2 - Bootstrapping
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{options}\NormalTok{( }\DataTypeTok{stringsAsFactors =} \OtherTok{FALSE}\NormalTok{ )}

\KeywordTok{cat}\NormalTok{( }\StringTok{"   we have data for"}\NormalTok{, n.ref, }\StringTok{"reference markers}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    we have data for 19 reference markers
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{cat}\NormalTok{( }\StringTok{"   we have data for"}\NormalTok{, n.species, }\StringTok{"species}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    we have data for 4 species
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{cat}\NormalTok{( }\StringTok{"   reference markers are"}\NormalTok{, marker.length, }\StringTok{"bases long}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    reference markers are 240 bases long
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.species, }\ControlFlowTok{function}\NormalTok{(i)\{}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(ref.hits)[i], }\StringTok{"has"}\NormalTok{, G.size[i],}\StringTok{"basepairs and"}\NormalTok{,N.reads[i],}\StringTok{"reads}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\} )}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## brown_bear has 2110508336 basepairs and 477378468 reads
## giant_panda has 2405352861 basepairs and 877225626 reads
## polar_bear has 2192934624 basepairs and 185147862 reads
## sun_bear has 2192934796 basepairs and 301081988 reads
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{### Point estimates of reference marker copy number in each species}
\NormalTok{### are based on read-counts and ALL reference markers for the specific species.}
\NormalTok{### The matrix ref.point contains one row for each species(__ rows)}
\NormalTok{### and one column for each marker (19), but since some}
\NormalTok{### markers have been excluded for the various species (see script1)}
\NormalTok{### there are some NA in the matrix.}

\CommentTok{# point estimation uses sample data to calculate a single value to serve as a best guess of copy numbers for sp.}
\NormalTok{ref.point <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\OtherTok{NA}\NormalTok{, }\DataTypeTok{nrow=}\KeywordTok{ncol}\NormalTok{(ref.hits), }\DataTypeTok{ncol=}\NormalTok{n.ref )}
\KeywordTok{rownames}\NormalTok{( ref.point ) <-}\StringTok{ }\KeywordTok{names}\NormalTok{( G.size )}
\KeywordTok{colnames}\NormalTok{( ref.point ) <-}\StringTok{ }\NormalTok{uref}
\ControlFlowTok{for}\NormalTok{( ss }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{n.species )\{}
  \KeywordTok{cat}\NormalTok{( }\KeywordTok{rownames}\NormalTok{(ref.point)[ss], }\StringTok{":}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
  \CommentTok{# same as C <- coverage[ss] }
\NormalTok{  C <-}\StringTok{ }\NormalTok{(N.reads[ss]}\OperatorTok{*}\NormalTok{marker.length)}\OperatorTok{/}\NormalTok{G.size[ss]         }\CommentTok{# coverage for this species}
\NormalTok{  idx.in <-}\StringTok{ }\KeywordTok{which}\NormalTok{( ref.keep[ss,] )                    }\CommentTok{# indices of ref. genes used for this species}
\NormalTok{  n.in <-}\StringTok{ }\KeywordTok{length}\NormalTok{( idx.in )                            }\CommentTok{# number of ref. genes used for this sp.}
  \CommentTok{# all observed hits for this sp., columns = ref.genes, rows = 4 cutoffs}
\NormalTok{  y.mat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( ref.hits[,ss], }\DataTypeTok{nrow=}\DecValTok{4}\NormalTok{, }\DataTypeTok{ncol=}\NormalTok{n.ref, }\DataTypeTok{byrow=}\NormalTok{F )}
\NormalTok{  y.mat <-}\StringTok{ }\NormalTok{y.mat[,idx.in]                             }\CommentTok{# hits for kept ref. genes for this sp.}
\NormalTok{  lst <-}\StringTok{ }\KeywordTok{copynum.iter.iter}\NormalTok{( y.mat, C, }\KeywordTok{rep}\NormalTok{( }\DecValTok{1}\NormalTok{, n.in ), }\DataTypeTok{as.integer=}\OtherTok{FALSE}\NormalTok{ )}
  \CommentTok{# estimated copy number for all kept ref. genes for this sp.}
\NormalTok{  ref.point[ss,idx.in] <-}\StringTok{ }\NormalTok{lst}\OperatorTok{$}\NormalTok{Gamma}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 5 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 6 
##  :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4 
##  :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4 
##  :
## copynum.iter.iter:
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 1 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 2 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 3 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 4 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 5 
## copynum.iter:
##  iteration 1 
##  iteration 2 
##  iteration.iteration 6
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{save}\NormalTok{( ref.point, }\DataTypeTok{file=}\StringTok{"res/ref_point.RData"}\NormalTok{ )}
\KeywordTok{boxplot}\NormalTok{(ref.point,}\DataTypeTok{las=}\DecValTok{2}\NormalTok{,}\DataTypeTok{pch=}\DecValTok{16}\NormalTok{,}\DataTypeTok{ylab=}\StringTok{"Reference marker copy number"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{full_CNV_pipeline_files/figure-latex/script2_ref_bootstraps_mammals-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{### The bootstrap procedure}

\CommentTok{# The difference to the point estimate above is that}
\CommentTok{# estimates are based on a bootstrap-sample of the reference}
\CommentTok{# markers for current species (not all markers), and this is repeated 1000 times.}
\NormalTok{### When estimating the copy number for marker g, this marker must of course}
\CommentTok{# be included, but the OTHER markers are bootstrapped 1000 times, and this}
\CommentTok{# is repeated for every marker for each species.}
\NormalTok{### The 3-dimensional array ref.boot stores the results.}

\CommentTok{# phi.hat - normalized average number of observed ref. gene hits for each cutoff for this sp.}
\CommentTok{# aka estimated stringency factor Fs with a hat (^)}
\NormalTok{phi.hat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\OperatorTok{-}\DecValTok{1}\NormalTok{, }\DataTypeTok{nrow=}\DecValTok{4}\NormalTok{, }\DataTypeTok{ncol=}\NormalTok{n.species )}
\KeywordTok{rownames}\NormalTok{( phi.hat ) <-}\StringTok{ }\KeywordTok{paste}\NormalTok{( }\StringTok{"log10(E)="}\NormalTok{, cuts, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{ ) }\CommentTok{# gives names to stringency rows }
\KeywordTok{colnames}\NormalTok{( phi.hat ) <-}\StringTok{ }\KeywordTok{colnames}\NormalTok{( ref.hits )}
\NormalTok{N.boot <-}\StringTok{ }\DecValTok{1000}
\CommentTok{# 3-dim array}
\NormalTok{ref.boot <-}\StringTok{ }\KeywordTok{array}\NormalTok{( }\OtherTok{NA}\NormalTok{, }\DataTypeTok{dim=}\KeywordTok{c}\NormalTok{(n.species, n.ref, N.boot), }\DataTypeTok{dimnames=}\KeywordTok{list}\NormalTok{( }\DataTypeTok{Species=}\KeywordTok{names}\NormalTok{(G.size), }\DataTypeTok{REF=}\NormalTok{uref, }\DataTypeTok{Boot=}\DecValTok{1}\OperatorTok{:}\NormalTok{N.boot ) )}
\ControlFlowTok{for}\NormalTok{( ss }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{n.species )\{}
  \KeywordTok{cat}\NormalTok{( }\KeywordTok{rownames}\NormalTok{(ref.point)[ss] )}
\NormalTok{  C <-}\StringTok{ }\NormalTok{(N.reads[ss]}\OperatorTok{*}\NormalTok{marker.length)}\OperatorTok{/}\NormalTok{G.size[ss]}
\NormalTok{  y.mat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( ref.hits[,ss], }\DataTypeTok{nrow=}\DecValTok{4}\NormalTok{, }\DataTypeTok{ncol=}\NormalTok{n.ref, }\DataTypeTok{byrow=}\NormalTok{F )}
\NormalTok{  idx.in <-}\StringTok{ }\KeywordTok{which}\NormalTok{( ref.keep[ss,] )}
\NormalTok{  y.mat <-}\StringTok{ }\NormalTok{y.mat[,idx.in]}
\NormalTok{  n.in <-}\StringTok{ }\KeywordTok{length}\NormalTok{( idx.in )}
  \ControlFlowTok{for}\NormalTok{( g }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{n.in )\{}
\NormalTok{    idb <-}\StringTok{ }\KeywordTok{which}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{n.in }\OperatorTok{!=}\StringTok{ }\NormalTok{g )    }\CommentTok{# randomly sampled g-1 genes from the ref.keep for this sp.}
    \ControlFlowTok{for}\NormalTok{( b }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{N.boot )\{}
      \CommentTok{# indices of sampled g-1 genes with replacement}
\NormalTok{      idx.boot <-}\StringTok{ }\KeywordTok{c}\NormalTok{( g, }\KeywordTok{sample}\NormalTok{( idb, }\DataTypeTok{size=}\NormalTok{(n.in}\OperatorTok{-}\DecValTok{1}\NormalTok{), }\DataTypeTok{replace=}\OtherTok{TRUE}\NormalTok{ ) ) }
\NormalTok{      y.boot <-}\StringTok{ }\NormalTok{y.mat[,idx.boot]    }\CommentTok{# randomized observed ref. hits regarding indices in idx.boot}
\NormalTok{      lst <-}\StringTok{ }\KeywordTok{copynum.iter.iter}\NormalTok{( y.boot, C, }\KeywordTok{rep}\NormalTok{( }\DecValTok{1}\NormalTok{, n.in ), }\DataTypeTok{max.iter=}\DecValTok{10}\NormalTok{, }\DataTypeTok{as.integer=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{verbose=}\OtherTok{FALSE}\NormalTok{ )}
\NormalTok{      ref.boot[ss,idx.in[g],b] <-}\StringTok{ }\NormalTok{lst}\OperatorTok{$}\NormalTok{Gamma[}\DecValTok{1}\NormalTok{]  }\CommentTok{# boot. estimate for this marker g for this sp.}
\NormalTok{    \}}
    \KeywordTok{cat}\NormalTok{( }\StringTok{"."}\NormalTok{ )}
\NormalTok{  \}}
  \KeywordTok{cat}\NormalTok{( }\StringTok{"}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ...............
## ...............
## ...............
## ................
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#save( ref.boot, file="res/ref_boot.RData" )}
\KeywordTok{boxplot}\NormalTok{(}\KeywordTok{t}\NormalTok{(ref.boot[}\DecValTok{3}\NormalTok{,,]), }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{4}\NormalTok{), }\DataTypeTok{las=}\DecValTok{2}\NormalTok{,}\DataTypeTok{ylab=}\StringTok{"Reference marker copy number"}\NormalTok{)}
\KeywordTok{abline}\NormalTok{(}\DataTypeTok{h=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{full_CNV_pipeline_files/figure-latex/script2_ref_bootstraps_mammals-2.pdf}

\subsection{Script 3 of the
procedure.}\label{script-3-of-the-procedure.}

Here we estimate the copy numbers of the MHC genes. Only
bootstrap-estimates are considered here. Point estimates are achieved by
averaging over the bootstrap-results for each gene and species.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{options}\NormalTok{( }\DataTypeTok{stringsAsFactors=}\NormalTok{F )}


\NormalTok{### Estimating copy number of MHC genes in each species}
\NormalTok{### based on read-counts and reference marker copy numbers.}
\NormalTok{### The bootstrapping is as follows:}
\NormalTok{### For each bootstrap sample, the corresponding bootstrap-result}
\NormalTok{### for the reference markers is used. In addition, not all}
\NormalTok{### reference markers are used, but another bootstrap-sample}
\NormalTok{### selects which reference markers to use each time.}

\NormalTok{bootstrap <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(mhc.hits, ref.keep) \{}
\NormalTok{  n.species =}\StringTok{ }\KeywordTok{ncol}\NormalTok{(mhc.hits)}
\NormalTok{  N.boot <-}\StringTok{ }\DecValTok{1000}
\NormalTok{  MHC.boot <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\DecValTok{0}\NormalTok{, }\DataTypeTok{nrow=}\NormalTok{n.species, }\DataTypeTok{ncol=}\NormalTok{N.boot )}
  \KeywordTok{rownames}\NormalTok{( MHC.boot ) <-}\StringTok{ }\KeywordTok{colnames}\NormalTok{( mhc.hits )}
  \ControlFlowTok{for}\NormalTok{( ss }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{n.species )\{}
    \KeywordTok{cat}\NormalTok{( }\StringTok{"Bootstrapping"}\NormalTok{, }\KeywordTok{rownames}\NormalTok{(MHC.boot)[ss], }\StringTok{"...}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{ )}
\NormalTok{    yg.mat <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( ref.hits[,ss], }\DataTypeTok{nrow=}\DecValTok{4}\NormalTok{, }\DataTypeTok{byrow=}\NormalTok{F )}
\NormalTok{    idx.in <-}\StringTok{ }\KeywordTok{which}\NormalTok{( ref.keep[ss,] )}
\NormalTok{    yg.mat <-}\StringTok{ }\NormalTok{yg.mat[,idx.in]}
\NormalTok{    GGG <-}\StringTok{ }\KeywordTok{length}\NormalTok{( idx.in )}
    \ControlFlowTok{for}\NormalTok{( b }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{N.boot )\{}
\NormalTok{      ref.gamma <-}\StringTok{ }\NormalTok{ref.boot[ss,idx.in,b]               }\CommentTok{# using the reference bootstrap results}
\NormalTok{      M <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{( }\OtherTok{NA}\NormalTok{, }\DataTypeTok{nrow=}\DecValTok{4}\NormalTok{, }\DataTypeTok{ncol=}\NormalTok{GGG )}
\NormalTok{      idx <-}\StringTok{ }\KeywordTok{sample}\NormalTok{( }\DecValTok{1}\OperatorTok{:}\NormalTok{GGG, }\DataTypeTok{size=}\NormalTok{GGG, }\DataTypeTok{replace=}\OtherTok{TRUE}\NormalTok{ )   }\CommentTok{# sampling which reference markers to use}
      \ControlFlowTok{for}\NormalTok{( i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{4}\NormalTok{ )\{}
        \ControlFlowTok{for}\NormalTok{( g }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\NormalTok{GGG )\{}
          \ControlFlowTok{if}\NormalTok{( (yg.mat[i,idx[g]] }\OperatorTok{!=}\StringTok{ }\DecValTok{0}\NormalTok{) }\OperatorTok{&}\StringTok{ }\NormalTok{(ref.gamma[idx[g]] }\OperatorTok{!=}\StringTok{ }\DecValTok{0}\NormalTok{) )\{}
\NormalTok{            M[i,g] <-}\StringTok{ }\NormalTok{ref.gamma[idx[g]] }\OperatorTok{*}\StringTok{ }\NormalTok{mhc.hits[i,ss]}\OperatorTok{/}\NormalTok{yg.mat[i,idx[g]] }
\NormalTok{          \}}
\NormalTok{        \}}
\NormalTok{      \}}
\NormalTok{      MHC.boot[ss,b] <-}\StringTok{ }\KeywordTok{mean}\NormalTok{( M, }\DataTypeTok{na.rm=}\OtherTok{TRUE}\NormalTok{ ) }\CommentTok{# averaging over all BLAST-cutoffs and selected references}
\NormalTok{    \}}
\NormalTok{  \}}
  
  \KeywordTok{return}\NormalTok{(MHC.boot)}
\NormalTok{\}}

\NormalTok{empty_mhc_hits <-}\StringTok{ }\KeywordTok{initialize_hits_table}\NormalTok{(all_species, }\DataTypeTok{gene_type =} \StringTok{"mhc"}\NormalTok{)}
\NormalTok{mhc_gene_names <-}\StringTok{ }\KeywordTok{get_gene_names}\NormalTok{(}\DataTypeTok{gene_type =} \StringTok{"mhc"}\NormalTok{)}
\NormalTok{all_mhc_hits <-}\StringTok{ }\KeywordTok{fill_hits}\NormalTok{(all_species, empty_mhc_hits, }\DataTypeTok{gene_type =} \StringTok{"mhc"}\NormalTok{)}

\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{length}\NormalTok{(mhc_gene_names)) \{}
\NormalTok{  mhc_gene_name <-}\StringTok{ }\NormalTok{mhc_gene_names[[i]]}
\NormalTok{  j <-}\StringTok{ }\NormalTok{i}\OperatorTok{*}\DecValTok{4} \OperatorTok{-}\StringTok{ }\DecValTok{3}
\NormalTok{  mhc.hits <-}\StringTok{ }\NormalTok{all_mhc_hits[j}\OperatorTok{:}\NormalTok{(j}\OperatorTok{+}\DecValTok{3}\NormalTok{),]}
\NormalTok{  MHC.boot <-}\StringTok{ }\KeywordTok{bootstrap}\NormalTok{(mhc.hits, ref.keep)}
  \KeywordTok{par}\NormalTok{( }\DataTypeTok{mar=}\KeywordTok{c}\NormalTok{(}\DecValTok{5}\NormalTok{,}\DecValTok{6}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{) )}
  \KeywordTok{boxplot}\NormalTok{( }\KeywordTok{t}\NormalTok{(MHC.boot ),}\DataTypeTok{cex.axis=}\FloatTok{0.6}\NormalTok{,}\DataTypeTok{horizontal=}\NormalTok{T,}\DataTypeTok{las=}\DecValTok{2}\NormalTok{,}\DataTypeTok{xlab=}\KeywordTok{paste}\NormalTok{(mhc_gene_name, }\StringTok{"copy number"}\NormalTok{))}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Bootstrapping brown_bear ...
## Bootstrapping giant_panda ...
## Bootstrapping polar_bear ...
## Bootstrapping sun_bear ...
\end{verbatim}

\includegraphics{full_CNV_pipeline_files/figure-latex/script3_MHC_copy_num_estimates-1.pdf}

\begin{verbatim}
## Bootstrapping brown_bear ...
## Bootstrapping giant_panda ...
## Bootstrapping polar_bear ...
## Bootstrapping sun_bear ...
\end{verbatim}

\includegraphics{full_CNV_pipeline_files/figure-latex/script3_MHC_copy_num_estimates-2.pdf}

\begin{verbatim}
## Bootstrapping brown_bear ...
## Bootstrapping giant_panda ...
## Bootstrapping polar_bear ...
## Bootstrapping sun_bear ...
\end{verbatim}

\includegraphics{full_CNV_pipeline_files/figure-latex/script3_MHC_copy_num_estimates-3.pdf}

\begin{verbatim}
## Bootstrapping brown_bear ...
## Bootstrapping giant_panda ...
## Bootstrapping polar_bear ...
## Bootstrapping sun_bear ...
\end{verbatim}

\includegraphics{full_CNV_pipeline_files/figure-latex/script3_MHC_copy_num_estimates-4.pdf}


\end{document}
